File: engines/py/fausth/deny_failure.py
from __future__ import annotations

from typing import Any

from .predicates import eval_predicate, get_path, is_missing


def _leaf_require_constraint(pred: dict[str, Any]) -> dict[str, Any]:
    for key in ("eq", "neq", "eq_path", "lt", "lte", "gt", "gte"):
        if key in pred:
            return {key: pred[key]}
    return {}


def failing_predicates(pred: dict[str, Any], snap: dict[str, Any]) -> list[dict[str, Any]]:
    if not isinstance(pred, dict):
        return []
    if "all" in pred:
        out: list[dict[str, Any]] = []
        for p in pred["all"]:
            out.extend(failing_predicates(p, snap))
        return out
    if "any" in pred:
        if eval_predicate(pred, snap):
            return []
        out = []
        for p in pred["any"]:
            out.extend(failing_predicates(p, snap))
        return out
    if "not" in pred:
        return [pred] if eval_predicate(pred["not"], snap) else []
    return [] if eval_predicate(pred, snap) else [pred]


def _lookup_unblock(
    key: str,
    eq_value: Any,
    checkpoints: list[dict[str, Any]],
) -> dict[str, Any] | None:
    for cp in checkpoints:
        if key in (cp.get("allow_set_keys") or []):
            return {"tool": cp["tool"], "set_key": key, "set_value": eq_value}
    return None


def build_predicate_failure(
    pred: dict[str, Any],
    snap: dict[str, Any],
    checkpoints: list[dict[str, Any]] | None = None,
) -> dict[str, Any]:
    failed: list[dict[str, Any]] = []
    for p in failing_predicates(pred, snap):
        if "path" not in p:
            continue
        path = str(p["path"])
        key = path[len("state.") :] if path.startswith("state.") else path
        current = get_path(snap, path)
        item: dict[str, Any] = {
            "path": path,
            "current": None if is_missing(current) else current,
            "require": _leaf_require_constraint(p),
        }
        if "eq" in p:
            unblock = _lookup_unblock(key, p["eq"], checkpoints or [])
            if unblock:
                item["unblock"] = unblock
        failed.append(item)
    return {"kind": "predicate", "failed": failed}


def build_missing_prior_tools_failure(missing: list[str]) -> dict[str, Any]:
    return {"kind": "missing_prior_tools", "missing_prior_tools": sorted(missing)}


def build_missing_prior_any_of_failure(options: list[str]) -> dict[str, Any]:
    return {"kind": "missing_prior_any_of", "options": sorted(options)}


def build_checkpoint_key_failure(key: str) -> dict[str, Any]:
    return {"kind": "checkpoint_key", "checkpoint_key": key}

File: engines/ts/src/deny-failure.ts
import { evalPredicate, getPath, isMissing } from "./predicates.js";
import type {
  CounterbalanceCheckpointPolicy,
  DenyFailure,
  DenyFailureItem,
  DenyFailureUnblock,
  Predicate,
  Snapshot,
} from "./types.js";

function leafRequireConstraint(p: Predicate): Record<string, unknown> {
  if ("eq" in p) return { eq: p.eq };
  if ("neq" in p) return { neq: p.neq };
  if ("eq_path" in p) return { eq_path: p.eq_path };
  if ("lt" in p) return { lt: p.lt };
  if ("lte" in p) return { lte: p.lte };
  if ("gt" in p) return { gt: p.gt };
  if ("gte" in p) return { gte: p.gte };
  return {};
}

export function failingPredicates(pred: Predicate, snap: Snapshot): Predicate[] {
  if ("all" in pred) {
    return pred.all.flatMap((p) => failingPredicates(p, snap));
  }
  if ("any" in pred) {
    if (evalPredicate(pred, snap)) return [];
    return pred.any.flatMap((p) => failingPredicates(p, snap));
  }
  if ("not" in pred) {
    return evalPredicate(pred.not, snap) ? [pred] : [];
  }
  return evalPredicate(pred, snap) ? [] : [pred];
}

function lookupUnblock(
  key: string,
  eqValue: unknown,
  checkpoints: CounterbalanceCheckpointPolicy[],
): DenyFailureUnblock | undefined {
  for (const cp of checkpoints) {
    if (cp.allow_set_keys.includes(key)) {
      return { tool: cp.tool, set_key: key, set_value: eqValue };
    }
  }
  return undefined;
}

export function buildPredicateFailure(
  pred: Predicate,
  snap: Snapshot,
  checkpoints: CounterbalanceCheckpointPolicy[] = [],
): DenyFailure {
  const failed: DenyFailureItem[] = [];
  for (const p of failingPredicates(pred, snap)) {
    if (!("path" in p)) continue;
    const path = p.path;
    const key = path.startsWith("state.") ? path.slice("state.".length) : path;
    const current = getPath(snap, path);
    const item: DenyFailureItem = {
      path,
      current: isMissing(current) ? null : current,
      require: leafRequireConstraint(p),
    };
    if ("eq" in p) {
      const unblock = lookupUnblock(key, p.eq, checkpoints);
      if (unblock) item.unblock = unblock;
    }
    failed.push(item);
  }
  return { kind: "predicate", failed };
}

export function buildMissingPriorToolsFailure(missing: string[]): DenyFailure {
  return { kind: "missing_prior_tools", missing_prior_tools: [...missing].sort() };
}

export function buildMissingPriorAnyOfFailure(options: string[]): DenyFailure {
  return { kind: "missing_prior_any_of", options: [...options].sort() };
}

export function buildCheckpointKeyFailure(key: string): DenyFailure {
  return { kind: "checkpoint_key", checkpoint_key: key };
}

/** Optional prose for live model tool results; not emitted in event logs. */
export function renderDenyFailureProse(failure: DenyFailure): string | null {
  if (failure.kind === "missing_prior_tools") {
    return `Call ${failure.missing_prior_tools.join(", ")} first.`;
  }
  if (failure.kind === "missing_prior_any_of") {
    return `Call one of: ${failure.options.join(", ")} first.`;
  }
  if (failure.kind === "checkpoint_key") {
    return `Checkpoint tool may not set '${failure.checkpoint_key}'.`;
  }
  const parts = failure.failed.map((f) => {
    const req = JSON.stringify(f.require);
    const cur = f.current === null ? "missing" : JSON.stringify(f.current);
    const unblock = f.unblock
      ? ` Call ${f.unblock.tool} with set ${JSON.stringify({ [f.unblock.set_key]: f.unblock.set_value })}.`
      : "";
    return `${f.path} requires ${req} (currently ${cur}).${unblock}`;
  });
  return parts.length > 0 ? parts.join(" ") : null;
}
